dockerspawner/systemuserspawner.py

```diff
@@ -191,8 +207,17 @@ def start(self, *, image=None, extra_create_kwargs=None, extra_host_config=None)
             self.extra_host_config.update(extra_host_config)
 
         self.extra_create_kwargs.setdefault('working_dir', self.homedir)
-        # systemuser image must be started as root
-        # relies on NB_UID and NB_USER handling in docker-stacks
-        self.extra_create_kwargs.setdefault('user', '0')
+        if self.run_as_root:
+            # systemuser images derived from Jupyter docker stacks
+            # can start as root and 'become' $NB_UID with proper
+            # name and everything.
+            # relies on $NB_UID and $NB_USER env handling in those images
+            # make this opt-in since without it, all users will run as root
+            self.extra_create_kwargs.setdefault('user', '0')
+        elif self.user_id >= 0:
+            user_s = str(self.user_id)
+            if self.group_id >= 0:
+                user_s = f"{user_s}:{self.group_id}"
+            self.extra_create_kwargs.setdefault('user', user_s)
```

Comment:
```
$ docker run --rm --user="0" --entrypoint sh jupyter/minimal-notebook -c 'echo $(id -u):$(id -g)'
0:0
$ docker run --rm --user="100" --entrypoint sh jupyter/minimal-notebook -c 'echo $(id -u):$(id -g)'
100:65534
$ docker run --rm --user="1000" --entrypoint sh jupyter/minimal-notebook -c 'echo $(id -u):$(id -g)'
1000:100
```

It seems like when group isn't explicitly set, it will get the user's associated group based on linux files influencing that (is it passwd?), and otherwise the nobody group `65534`.

Would you say the logic is fine about this? With a non-root user and non-root group, the jupyter/docker-stacks script doing fancy stuff with the username and such won't work, so my question is more concretely - are you assuming here it would work when group_id isn't specified?

Hmmm... I think there is something wrong with this logic no matter what as if group_id is explicitly set to zero, it would still be ignored if someone set the user to `100` for example with the logic `self.group_id >= 0`.

Reply:
I don't think docker allows setting group but not user:

```
  -u, --user string                    Username or UID (format: <name|uid>[:<group|gid>])
```

So it's not valid to have group set but user not set. The question then becomes: do we raise informatively, or not. I went with adding a warning for this case.

Reply:
> I think there is something wrong with this logic no matter what as if group_id is explicitly set to zero

No, gid=0 is respected because it's `>= 0` not `> 0`.

Reply:
Eeeeh woops, sorry for the noise!